kobo-uncaged/kunc/kunc.go

```diff
@@ -0,0 +1,296 @@
+// A nil slice is interpreted has having no books on the device
+func (ku *koboUncaged) GetDeviceBookList() []uc.BookCountDetails {
+	bc := []uc.BookCountDetails{}
+	for _, md := range ku.k.MetadataMap {
+		lastMod := time.Now()
+		if md.LastModified != nil {
+			lastMod, _ = time.Parse(time.RFC3339, *md.LastModified)
+		}
+		bcd := uc.BookCountDetails{
+			UUID:         md.UUID,
+			Lpath:        md.Lpath,
+			LastModified: lastMod,
+		}
+		bcd.Extension = filepath.Ext(md.Lpath)
+		bc = append(bc, bcd)
+	}
+	//spew.Dump(bc)
+	return bc
+}
+
+// GetMetadataList sends complete metadata for the books listed in lpaths, or for
+// all books on device if lpaths is empty
+func (ku *koboUncaged) GetMetadataList(books []uc.BookID) []map[string]interface{} {
+	//spew.Dump(ku.k.MetadataMap)
+	//spew.Dump(books)
+	mdList := []map[string]interface{}{}
+	if len(books) > 0 {
+		for _, bk := range books {
+			cid := util.LpathToContentID(bk.Lpath, string(ku.k.ContentIDprefix))
+			fmt.Println(cid)
+			md := map[string]interface{}{}
+			//spew.Dump(ku.k.MetadataMap[cid])
+			mapstructure.Decode(ku.k.MetadataMap[cid], &md)
+			mdList = append(mdList, md)
+		}
+	} else {
+		for _, kmd := range ku.k.MetadataMap {
+			md := map[string]interface{}{}
+			//spew.Dump(kmd)
+			mapstructure.Decode(kmd, &md)
+			mdList = append(mdList, md)
+		}
+	}
+	return mdList
+}
+
+// GetDeviceInfo asks the client for information about the drive info to use
+func (ku *koboUncaged) GetDeviceInfo() uc.DeviceInfo {
+	return ku.k.DriveInfo
+}
+
+// SetDeviceInfo sets the new device info, as comes from calibre. Only the nested
+// struct DevInfo is modified.
+func (ku *koboUncaged) SetDeviceInfo(devInfo uc.DeviceInfo) {
+	ku.k.DriveInfo = devInfo
+	ku.k.SaveDeviceInfo()
+}
+
+// UpdateMetadata instructs the client to update their metadata according to the
+// new slice of metadata maps
+func (ku *koboUncaged) UpdateMetadata(mdList []map[string]interface{}) {
+	for _, md := range mdList {
+		koboMD := device.CreateKoboMetadata()
+		mapstructure.Decode(md, &koboMD)
+		koboMD.Thumbnail = nil
+		cid := util.LpathToContentID(koboMD.Lpath, string(ku.k.ContentIDprefix))
+		ku.k.MetadataMap[cid] = koboMD
+		ku.k.UpdatedMetadata = append(ku.k.UpdatedMetadata, cid)
+	}
+	ku.k.WriteMDfile()
+	ku.k.WriteUpdateMDfile()
+}
+
+// GetPassword gets a password from the user.
+func (ku *koboUncaged) GetPassword(calibreInfo uc.CalibreInitInfo) string {
+	return ku.k.Passwords.NextPassword()
+}
+
+// GetFreeSpace reports the amount of free storage space to Calibre
+func (ku *koboUncaged) GetFreeSpace() uint64 {
+	// Note, this method of getting available disk space is Linux specific...
+	// Don't try to run this code on Windows. It will probably fall over
+	var fs syscall.Statfs_t
+	err := syscall.Statfs(ku.k.BKRootDir, &fs)
+	if err != nil {
+		log.Println(err)
+		// Despite the error, we return an arbitrary amount. Thoughts on this?
```

Comment:
I would agree with this, as the [errors returned by statfs](http://man7.org/linux/man-pages/man2/statfs.2.html) aren't really fixable in the code or fatal. The ones that are will cause errors soon enough anyways.

Reply:
Thanks. Is what I thought